### PR TITLE
Add rule management component.

### PR DESCRIPTION
**Pull request by malithie**
### Proposed changes in this pull request

Resolves https://github.com/wso2/product-is/issues/22026

This PR adds the rule management service component.

### CODE REVIEW

**Comment by malithie** [comment]
> ## [Quality Gate Passed](https://sonarcloud.io/dashboard?id=wso2_carbon-identity-framework&pullRequest=6204) **Quality Gate passed**
> Issues  [7 New issues](https://sonarcloud.io/project/issues?id=wso2_carbon-identity-framework&pullRequest=6204&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true)  [0 Accepted issues](https://sonarcloud.io/project/issues?id=wso2_carbon-identity-framework&pullRequest=6204&issueStatuses=ACCEPTED)
> 
> Measures  [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=wso2_carbon-identity-framework&pullRequest=6204&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true)  [0.0% Coverage on New Code](https://sonarcloud.io/component_measures?id=wso2_carbon-identity-framework&pullRequest=6204&metric=new_coverage&view=list)  [0.0% Duplication on New Code](https://sonarcloud.io/component_measures?id=wso2_carbon-identity-framework&pullRequest=6204&metric=new_duplicated_lines_density&view=list)
> 
> [See analysis details on SonarQube Cloud](https://sonarcloud.io/dashboard?id=wso2_carbon-identity-framework&pullRequest=6204)

Applicable suggestions are addressed

File: components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/RuleManagementDAOImpl.java [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+
+    private void addRuleToDB(Rule rule, int tenantId) throws TransactionException, RuleManagementServerException {
+
+        InputStream ruleJson = convertRuleToJson(rule);
+        NamedJdbcTemplate jdbcTemplate = new NamedJdbcTemplate(IdentityDatabaseUtil.getDataSource());
+        jdbcTemplate.withTransaction(template -> {
+            template.executeInsert(RuleSQLConstants.Query.ADD_RULE,
+                    statement -> {
+                        statement.setString(RuleSQLConstants.Column.UUID, rule.getId());
+                        statement.setBlob(RuleSQLConstants.Column.RULE_INDEX, ruleJson);
```
Comment (ashanthamara) [inline components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/RuleManagementDAOImpl.java:204]: Let's use setBinaryStream(String name, InputStream inputStream, long length) method with the column name reference.
https://github.com/wso2/carbon-utils/blob/90b5d4300fafc5d4abaaadb7e41a9a340429f390/components/org.wso2.carbon.database.utils/src/main/java/org/wso2/carbon/database/utils/jdbc/NamedPreparedStatement.java#L226

File: components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/RuleManagementDAO.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+/**
+ * Rule Management DAO.
+ * This class is used to perform CRUD operations on Rule in the datastore.
+ */
+public interface RuleManagementDAO {
+
+    /**
+     * Add a new Rule.
+     * @param rule Rule object
```
Comment (ashanthamara) [inline components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/RuleManagementDAO.java:32]: Check formatting.
Add a new line between the description and params.

Check other places as well

File: components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/CacheBackedRuleManagementDAO.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        ruleManagementDAO.updateRule(rule, tenantId);
+    }
+
+    /**
+     * Delete a Rule.
+     * This method clears the cache entry upon rule deletion.
+     *
+     * @param ruleId   Rule ID
+     * @param tenantId Tenant ID
+     */
```
Comment (ashanthamara) [inline components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/CacheBackedRuleManagementDAO.java:82]: Missing @throws line. 
Check other places

File: components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/RuleManagementDAOImpl.java [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+     *
+     * @param rule     Rule object
+     * @param tenantId Tenant ID
+     * @throws RuleManagementException If an error occurs while adding the rule to the database.
+     */
+    @Override
+    public void addRule(Rule rule, int tenantId) throws RuleManagementException {
+
+        try {
+            addRuleToDB(rule, tenantId);
```
Comment (ashanthamara) [inline components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/RuleManagementDAOImpl.java:58]: If we want to following tasks in one transaction, then we need to initiate the jdbcTemplate.withTransaction() from this level

File: components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/RuleManagementDAOImpl.java [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+     *
+     * @param rule     Rule object
+     * @param tenantId Tenant ID
+     * @throws RuleManagementException If an error occurs while updating the rule in the database.
+     */
+    @Override
+    public void updateRule(Rule rule, int tenantId) throws RuleManagementException {
+
+        try {
+            updateRuleInDB(rule, tenantId);
```
Comment (ashanthamara) [inline components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/RuleManagementDAOImpl.java:77]: Same as above comment

File: features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/dbscripts/h2.sql [hunk 6]
```diff
@@ -1380,6 +1380,24 @@ CREATE TABLE IF NOT EXISTS IDN_OAUTH2_TOKEN_CLAIMS (
...
+CREATE TABLE IF NOT EXISTS IDN_RULE (
+	UUID CHAR(36) NOT NULL,
+	RULE BLOB NOT NULL,
+	IS_ACTIVE BOOLEAN DEFAULT TRUE,
+	TENANT_ID INTEGER NOT NULL,
+	PRIMARY KEY (UUID)
+);
+
+CREATE TABLE IF NOT EXISTS IDN_RULE_REFERENCES (
+    ID INTEGER NOT NULL AUTO_INCREMENT,
```
Comment (darshanasbg) [inline features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/dbscripts/h2.sql:1392]: Fix formatting

File: components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/RuleManagementDAOImpl.java (hunk 4 above)
Comment (malithie) [reply components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/RuleManagementDAOImpl.java]: There seems to be some issue with the current utility that handles transactions. When data is not committed if the rest of the updates are followed on top of uncommitted data they seem to fail.
https://github.com/wso2-enterprise/asgardeo-product/issues/27910

We have to look at this separately. So for now for rule addition transaction flow will be avoided. More over, looks like we haven't seem to have tried this transactional flow in similar scenarios anywhere in the product. I couldn't locate any even though I looked for references

File: features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/dbscripts/h2.sql [hunk 7]
```diff
@@ -1501,3 +1522,8 @@ CREATE INDEX IDX_IDN_ACTION_ENDPOINT_AU_TI ON IDN_ACTION_ENDPOINT (ACTION_UUID,
 -- CERTIFICATE --
 CREATE INDEX IDX_IDN_CERTIFICATE_ID_TID ON IDN_CERTIFICATE (ID, TENANT_ID);
 CREATE INDEX IDX_IDN_CERTIFICATE_UUID_TID ON IDN_CERTIFICATE (UUID, TENANT_ID);
+
+-- RULES --
+CREATE INDEX IDX_IDN_RULE_UUID_TID ON IDN_RULE (UUID, TENANT_ID);
+CREATE INDEX IDX_IDN_RULE_REF_RID_TID ON IDN_RULE_REFERENCES (RULE_ID, TENANT_ID);
+CREATE INDEX IDX_IDN_RULE_REF_FN_FREF_TID ON IDN_RULE_REFERENCES (FIELD_NAME, FIELD_REFERENCE, TENANT_ID);
```
Comment (darshanasbg) [inline features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/dbscripts/h2.sql:1529]: Why we need this index?

File: components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/RuleManagementDAOImpl.java [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+                        statement.setBinaryStream(RuleSQLConstants.Column.RULE, ruleJsonAsInputStream,
+                                ruleJsonStreamLength);
+                        statement.setBoolean(RuleSQLConstants.Column.IS_ACTIVE, rule.isActive());
+                        statement.setInt(RuleSQLConstants.Column.TENANT_ID, tenantId);
+                        statement.setString(RuleSQLConstants.Column.VERSION, V1);
+                    }, rule, true));
+            return null;
+        });
+
+        return internalId.get();
```
Comment (ashanthamara) [inline components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/RuleManagementDAOImpl.java:231]: Some db types doesn't support returning the auto generated id. In such cases we may need to check whether the value of `internalId` is zero and then use a different way to fetch the id (Eg: getRuleByUUID)

File: features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/dbscripts/h2.sql (hunk 7 above)
Comment (malithie) [reply features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/dbscripts/h2.sql:1529]: This index is needed as we need to find rules based on fields and field values once resources associated for those fields and filed values can get deleted (e.g:, an application that is referred in a rule gets deleted).

However, at this point of time even though the index is present the query or the respective funtionality that would utilize this is not present.
So let's add it back when needed

File: components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/CacheBackedRuleManagementDAO.java (hunk 3 above)
Comment (malithie) [reply components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/CacheBackedRuleManagementDAO.java:85]: I purposefully removed this exception initially as this method contract is overall explained and what's important to mention is about what this particular layer implements.

Anyhow for consistency added back and fixed

File: components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/RuleManagementDAOImpl.java (hunk 4 above)
Comment (malithie) [reply components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/RuleManagementDAOImpl.java]: As discussed adding 'withTransaction' for every query started using the same connection and resolved this issue. Thus this is updated accordingly

File: features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/dbscripts/h2.sql (hunk 7 above)
Comment (malithie) [reply features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/dbscripts/h2.sql]: Removed the redundant index at this point of time based on the discussion above from commit: https://github.com/wso2/carbon-identity-framework/pull/6204/commits/c904b60008bc3efb85b9a7f46790b3c380649924

File: components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/RuleManagementDAOImpl.java [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+                        statement -> {
+                            statement.setBoolean(RuleSQLConstants.Column.IS_ACTIVE, false);
+                            statement.setString(RuleSQLConstants.Column.RULE_EXTERNAL_ID, ruleId);
+                            statement.setInt(RuleSQLConstants.Column.TENANT_ID, tenantId);
+                        });
+                return null;
+            });
+        } catch (TransactionException e) {
+            throw new RuleManagementServerException("Error while deactivating the rule in the system.", e);
+        }
```
Comment (ashanthamara) [inline components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/RuleManagementDAOImpl.java:212]: Shall we have a common private method for activating and deactivating rule ? We can pass the status as an argument.

File: components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/RuleManagementDAOImpl.java [hunk 10]
```diff
@@ -0,0 +1,328 @@
...
+                    statement -> {
+                        statement.setString(RuleSQLConstants.Column.RULE_EXTERNAL_ID, rule.getId());
+                        statement.setBinaryStream(RuleSQLConstants.Column.RULE, ruleJsonAsInputStream,
+                                ruleJsonStreamLength);
+                        statement.setBoolean(RuleSQLConstants.Column.IS_ACTIVE, rule.isActive());
+                        statement.setInt(RuleSQLConstants.Column.TENANT_ID, tenantId);
+                        statement.setString(RuleSQLConstants.Column.VERSION, V1);
+                    }, rule, true));
+            return null;
+        });
```
Comment (ashanthamara) [inline components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/RuleManagementDAOImpl.java:233]:
```suggestion
        AtomicInteger internalId = new AtomicInteger();
        int internalId = jdbcTemplate.withTransaction(template -> template.executeInsert(RuleSQLConstants.Query.ADD_RULE,
                    statement -> {
                        statement.setString(RuleSQLConstants.Column.RULE_EXTERNAL_ID, rule.getId());
                        statement.setBinaryStream(RuleSQLConstants.Column.RULE, ruleJsonAsInputStream,
                                ruleJsonStreamLength);
                        statement.setBoolean(RuleSQLConstants.Column.IS_ACTIVE, rule.isActive());
                        statement.setInt(RuleSQLConstants.Column.TENANT_ID, tenantId);
                        statement.setString(RuleSQLConstants.Column.VERSION, V1);
                    }, rule, true)));
```

File: components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/RuleManagementDAOImpl.java [hunk 11]
```diff
@@ -0,0 +1,328 @@
...
+        jdbcTemplate.withTransaction(template -> {
+            template.executeUpdate(RuleSQLConstants.Query.UPDATE_RULE,
+                    statement -> {
+                        statement.setBinaryStream(RuleSQLConstants.Column.RULE, ruleJsonAsInputStream,
+                                ruleJsonStreamLength);
+                        statement.setString(RuleSQLConstants.Column.RULE_EXTERNAL_ID, rule.getId());
+                        statement.setInt(RuleSQLConstants.Column.TENANT_ID, tenantId);
+                    });
+            return null;
+        });
```
Comment (ashanthamara) [inline components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/RuleManagementDAOImpl.java:293]:
```suggestion
        jdbcTemplate.withTransaction(template -> 
            template.executeUpdate(RuleSQLConstants.Query.UPDATE_RULE,
                    statement -> {
                        statement.setBinaryStream(RuleSQLConstants.Column.RULE, ruleJsonAsInputStream,
                                ruleJsonStreamLength);
                        statement.setString(RuleSQLConstants.Column.RULE_EXTERNAL_ID, rule.getId());
                        statement.setInt(RuleSQLConstants.Column.TENANT_ID, tenantId);
                    }));
```

File: components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/RuleManagementDAOImpl.java [hunk 12]
```diff
@@ -0,0 +1,328 @@
...
+
+        NamedJdbcTemplate jdbcTemplate = new NamedJdbcTemplate(IdentityDatabaseUtil.getDataSource());
+        jdbcTemplate.withTransaction(template -> {
+            template.executeUpdate(RuleSQLConstants.Query.DELETE_RULE_REFERENCES,
+                    statement -> {
+                        statement.setInt(RuleSQLConstants.Column.RULE_REFERENCE_ID, internalRuleId);
+                        statement.setInt(RuleSQLConstants.Column.TENANT_ID, tenantId);
+                    });
+            return null;
+        });
```
Comment (ashanthamara) [inline components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/RuleManagementDAOImpl.java:306]:
```suggestion
        jdbcTemplate.withTransaction(template -> 
            template.executeUpdate(RuleSQLConstants.Query.DELETE_RULE_REFERENCES,
                    statement -> {
                        statement.setInt(RuleSQLConstants.Column.RULE_REFERENCE_ID, internalRuleId);
                        statement.setInt(RuleSQLConstants.Column.TENANT_ID, tenantId);
                    }));
```

File: components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/RuleManagementDAOImpl.java [hunk 13]
```diff
@@ -0,0 +1,328 @@
...
+                                    statement.setString(RuleSQLConstants.Column.FIELD_REFERENCE,
+                                            expression.getValue().getFieldValue());
+                                    statement.setInt(RuleSQLConstants.Column.TENANT_ID, tenantId);
+                                    statement.addBatch();
+                                }
+                            }
+                        }
+                    }, null);
+            return null;
+        });
```
Comment (ashanthamara) [inline components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/RuleManagementDAOImpl.java:275]:
```suggestion
        jdbcTemplate.withTransaction(template -> 
            template.executeBatchInsert(RuleSQLConstants.Query.ADD_RULE_REFERENCES,
                    statement -> {
                        for (ANDCombinedRule rule1 : orCombinedRule.getRules()) {
                            for (Expression expression : rule1.getExpressions()) {
                                if (expression.getValue().getType() == Value.Type.REFERENCE) {
                                    statement.setInt(RuleSQLConstants.Column.RULE_REFERENCE_ID, internalRuleId);
                                    statement.setString(RuleSQLConstants.Column.FIELD_NAME, expression.getField());
                                    statement.setString(RuleSQLConstants.Column.FIELD_REFERENCE,
                                            expression.getValue().getFieldValue());
                                    statement.setInt(RuleSQLConstants.Column.TENANT_ID, tenantId);
                                    statement.addBatch();
                                }
                            }
                        }
                    }, null));
```

File: components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/RuleManagementDAOImpl.java (hunk 9 above)
Comment (malithie) [reply components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/RuleManagementDAOImpl.java:212]: I don't really see a necessity here.

**Comment by malithie** [comment]
> 9 New issues

All applicable suggestions are addressed

File: components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/RuleManagementDAOImpl.java (hunk 8 above)
Comment (malithie) [reply components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/RuleManagementDAOImpl.java]: Fixed

**Review by Thisara-Welmilla — DISMISSED** [review]

File: components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/CacheBackedRuleManagementDAO.java [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+     *
+     * @param ruleId   Rule ID
+     * @param tenantId Tenant ID
+     * @return Rule object
+     * @throws RuleManagementException Rule Management Exception
+     */
+    @Override
+    public Rule getRuleByRuleId(String ruleId, int tenantId) throws RuleManagementException {
+
+        RuleCacheEntry ruleCacheEntry = RuleCache.getInstance().getValueFromCache(new RuleCacheKey(ruleId), tenantId);
```
Comment (ashanthamara) [inline components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/CacheBackedRuleManagementDAO.java:107]:
```suggestion
        RuleCacheEntry ruleCacheEntry = ruleCache.getValueFromCache(new RuleCacheKey(ruleId), tenantId);
```

File: components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/CacheBackedRuleManagementDAO.java [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+        if (ruleCacheEntry != null && ruleCacheEntry.getRule() != null) {
+            LOG.debug("Rule cache hit for rule id: " + ruleId + ". Returning from cache.");
+            return ruleCacheEntry.getRule();
+        }
+
+        Rule rule = ruleManagementDAO.getRuleByRuleId(ruleId, tenantId);
+        if (rule != null) {
+            LOG.debug("Rule cache miss for rule id: " + ruleId + ". Adding to cache.");
+            ruleCache.addToCache(new RuleCacheKey(ruleId), new RuleCacheEntry(rule), tenantId);
+        }
```
Comment (ashanthamara) [inline components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/CacheBackedRuleManagementDAO.java:117]: Shall we add else part and add the debug log,
LOG.debug("Rule for rule id:  " + ruleId + " not found in cache or DB.");

File: components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/service/impl/RuleManagementServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+     * @param tenantDomain Tenant domain.
+     * @return Added rule.
+     * @throws RuleManagementException If an error occurs while adding the rule.
+     */
+    @Override
+    public Rule addRule(Rule rule, String tenantDomain) throws RuleManagementException {
+
+        int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
+        ruleManagementDAO.addRule(rule, tenantId);
+        return ruleManagementDAO.getRuleByRuleId(rule.getId(), tenantId);
```
Comment (ashanthamara) [inline components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/service/impl/RuleManagementServiceImpl.java:61]:
```suggestion
        return getRuleByRuleId(rule.getId(), tenantId);
```
Shall we replace ruleManagementDAO.getRuleByRuleId() with the public method ?

Comment (malithie) [reply components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/service/impl/RuleManagementServiceImpl.java:61]: It's not necessary. I rather like to access the dao layer. service layer api may have additional logics which is not quite related

File: components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/CacheBackedRuleManagementDAO.java (hunk 15 above)
Comment (malithie) [reply components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/dao/impl/CacheBackedRuleManagementDAO.java:117]: Well I don't think it's necessary. This indicates that it's a cache miss if the entry is already available in the DB.
This layer is not responsible on the entry not being available in DB IMO, so such log is redundant and is just some info that eventually gets known to the consumer of this API

**Review by ashanthamara — APPROVED** [review]

**Pull request merged**